components/function-controller/pkg/configwatcher/credentials_service.go

```diff
@@ -0,0 +1,172 @@
+	}
+	if list == nil || len(list.Items) == 0 {
+		return errors.New(fmt.Sprintf("not found Base Credentials in '%s' namespace by labelSelector '%s'", s.config.BaseNamespace, labelSelector))
+	}
+
+	if s.cachedCredentials == nil {
+		s.cachedCredentials = make(map[string]*corev1.Secret)
+	}
+
+	for i, credential := range list.Items {
+		credentialType := credential.Labels[CredentialsLabel]
+		if credentialType != "" {
+			s.cachedCredentials[credentialType] = &list.Items[i]
+		}
+	}
+	return nil
+}
+
+func (s *CredentialsService) UpdateCachedCredential(credential *corev1.Secret) error {
+	if credential == nil {
+		return errors.New("Credential is nil")
+	}
+
+	credentialType := credential.Labels[CredentialsLabel]
+	if credentialType == "" {
+		return errors.New(fmt.Sprintf("Credential %v hasn't '%s' label", credential, CredentialsLabel))
+	}
+
+	if s.cachedCredentials == nil {
+		err := s.UpdateCachedCredentials()
+		if err != nil {
+			return err
+		}
+	}
+
+	s.cachedCredentials[credentialType] = credential
+	return nil
+}
+
+func (s *CredentialsService) HandleCredentialsInNamespace(namespace string) error {
+	credentials, err := s.GetCredentials()
+	if err != nil {
+		return errors.Wrapf(err, "while handling Credentials in '%s' namespace", namespace)
+	}
+
+	for _, credential := range credentials {
+		err := s.updateCredentialInNamespace(credential, namespace)
+		if err != nil {
+			return errors.Wrapf(err, "while handling Credentials in '%s' namespace", namespace)
+		}
+	}
+
+	return nil
+}
+
+func (s *CredentialsService) HandleCredentialInNamespaces(credential *corev1.Secret, namespaces []string) error {
+	for _, namespace := range namespaces {
+		err := s.updateCredentialInNamespace(credential, namespace)
+		if err != nil {
+			return errors.Wrapf(err, "while handling Credential '%s' in %v namespaces", credential.Name, namespaces)
+		}
+	}
+	return nil
+}
+
+func (s *CredentialsService) IsBaseCredential(credential *corev1.Secret) bool {
+	return credential.Namespace == s.config.BaseNamespace && credential.Labels[ConfigLabel] == CredentialsLabelValue
+}
+
+func (s *CredentialsService) createCredentialInNamespace(credential *corev1.Secret, namespace string) error {
+	newCredential := s.copyCredentials(credential, namespace)
+	_, err := s.coreClient.Secrets(namespace).Create(newCredential)
+	if err != nil {
+		if apiErrors.IsAlreadyExists(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "while creating Credential '%s' in '%s' namespace", newCredential.Name, namespace)
+	}
+
+	return nil
+}
+
+func (s *CredentialsService) updateCredentialInNamespace(credential *corev1.Secret, namespace string) error {
+	newCredential := s.copyCredentials(credential, namespace)
+	_, err := s.coreClient.Secrets(namespace).Update(newCredential)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			err = s.createCredentialInNamespace(credential, namespace)
+			if err != nil {
+				return err
+			}
+		} else {
```

Comment:
`else` is not necessary

Reply:
`else` is necessary, because if resource is not found, then it try create this resource. `else` skips error if resource creation succeeds

Reply:
Please try to understand the flow :)